content/concurrency.md

```diff
@@ -1,14 +1,57 @@
+Web apps have taken a variety of process-execution forms. For example,
+PHP processes run as child processes of Apache, started on demand as needed
+by request volume. Java processes take the opposite approach, with the JVM
+providing one massive uberprocess that reserves a large block of system
+resources (CPU and memory) on startup, with concurrency managed internally
+via threads.
 
-The process model truly shines when it comes time to scale out.  The [share-nothing, horizontally partitionable nature of twelve-factor app processes](./processes) means that adding more concurrency is a simple and reliable operation.  The array of process types and number of processes of each type is known as the *process formation*.
+This does not exclude individual processes from handling their own internal
+multiplexing, via threads inside the runtime VM, or the async/evented model
+found in tools such as [EventMachine](https://github.com/eventmachine/eventmachine),
+[Twisted](http://twistedmatrix.com/trac/), or
+[Node.js](http://nodejs.org/).
 
-Twelve-factor app processes [should never daemonize](http://dustin.github.com/2010/02/28/running-processes.html) or write PID files.  Instead, rely on the operating system's process manager (such as [systemd](https://www.freedesktop.org/wiki/Software/systemd/), a distributed process manager on a cloud platform, or a tool like [Foreman](http://blog.daviddollar.org/2011/05/06/introducing-foreman.html) in development) to manage [output streams](./logs), respond to crashed processes, and handle user-initiated restarts and shutdowns.
+##### Guidance
+
+An individual VM can only grow so large (vertical scale), so the application
+must also be able to span multiple processes running on multiple physical or
+virtual machines.
+
+#### 2\. A twelve-factor app scales out by assigning workloads to separate process types.
+
+The process model truly shines when it comes time to scale out.
+The [share-nothing, horizontally partitionable nature of twelve-factor app processes](./processes)
+means that adding more concurrency is a simple and reliable operation.
+The array of process types and number of processes of each type is known as
+the *process formation*.
+
+#### 3\. A twelve-factor app relies on the execution environment to manage processes.
+
+Twelve-factor app processes [should never daemonize](http://dustin.github.com/2010/02/28/running-processes.html)
+or write PID files. Instead, they depend on the execution environment for
+controlling their lifecycle.
+
+##### Guidance
+
+Rely on the operating system's process manager (such as
+[systemd](https://www.freedesktop.org/wiki/Software/systemd/), a distributed
+process manager on a cloud platform, or a tool like
+[Foreman](http://blog.daviddollar.org/2011/05/06/introducing-foreman.html)
+in development. These managers handle
+[output streams](./logs.md), respond to crashed processes, and manage restarts
+and shutdowns.
```

Comment:
This should probably be updated to reference Docker as well.

Reply:
@evankanderson  I wholeheartedly agree with most of these perspectives but the scope of this pr is limited to #13 . It is supposed to just reorganize while changing the language as little as possible. I want to get to that part of the discussion asap since it is the fun part. Not sure how to merge this without losing the excellent discussion though

Reply:
All good, I captured the discussion so we can pick up one these points during the next rev.